README.md

# PureNES

[![build](https://github.com/zeeps31/purenes/actions/workflows/build.yml/badge.svg)](https://github.com/zeeps31/purenes/actions/workflows/build.yml)
[![codecov](https://codecov.io/gh/zeeps31/purenes/branch/main/graph/badge.svg?token=BJ3OVYA7GN)](https://codecov.io/gh/zeeps31/purenes)
[![docs](https://github.com/zeeps31/purenes/actions/workflows/docs.yml/badge.svg)](https://github.com/zeeps31/purenes/actions/workflows/docs.yml)

## Overview 

PureNES is a "Pure Python" NES emulator.


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.33.0 (2022-07-19)
### Feature
* **cpu:** Implement NOP operation ([#134](https://github.com/zeeps31/purenes/issues/134)) ([`442e7fa`](https://github.com/zeeps31/purenes/commit/442e7faabcd5cb197130371850c3bcb4a246081f))
* **cpu:** Implement RTI operation ([#133](https://github.com/zeeps31/purenes/issues/133)) ([`9de45fc`](https://github.com/zeeps31/purenes/commit/9de45fcc18c329ac3ee147a8a07b017c66341253))
* **cpu:** Implement RTS operation ([#132](https://github.com/zeeps31/purenes/issues/132)) ([`dd43ab0`](https://github.com/zeeps31/purenes/commit/dd43ab079f3c10d617c87bd74593cd62bae72a10))
* **cpu:** Implement comparison operations ([#131](https://github.com/zeeps31/purenes/issues/131)) ([`fca5dd5`](https://github.com/zeeps31/purenes/commit/fca5dd505ba94286b10610dd8b10377701b50160))
* **cpu:** Implement ROR operation ([#130](https://github.com/zeeps31/purenes/issues/130)) ([`de03d3a`](https://github.com/zeeps31/purenes/commit/de03d3a2fa380ef2e6130217891f74c1dcde9bf0))
* **cpu:** Implement LSR operation ([#129](https://github.com/zeeps31/purenes/issues/129)) ([`464df2d`](https://github.com/zeeps31/purenes/commit/464df2d6970a30bf18ec5a991333d6316dd2eb92))
* **cpu:** Implement EOR operation ([#128](https://github.com/zeeps31/purenes/issues/128)) ([`6663cfa`](https://github.com/zeeps31/purenes/commit/6663cfaca3ead5cbd60b1bbc6fe45bf154c7022a))
* **cpu:** Implement SBC operation ([#126](https://github.com/zeeps31/purenes/issues/126)) ([`86171bf`](https://github.com/zeeps31/purenes/commit/86171bf3a229967cf2204e0910c4a5cb83388bb8))
* **cpu:** Implement ADC operation ([#125](https://github.com/zeeps31/purenes/issues/125)) ([`62dcfa3`](https://github.com/zeeps31/purenes/commit/62dcfa3c4b74523e8229990f4a2ec0c38868cff6))
* **cpu:** Implement increment and decrement register operations ([#123](https://github.com/zeeps31/purenes/issues/123)) ([`fbcdd2c`](https://github.com/zeeps31/purenes/commit/fbcdd2c034e7a153dd5aa757d3d2551d6894a765))
* **cpu:** Implement INC operation ([#122](https://github.com/zeeps31/purenes/issues/122)) ([`2ce1fb7`](https://github.com/zeeps31/purenes/commit/2ce1fb7467580edcd62258f1e389c0e479af6852))
* **cpu:** Implement DEC operation ([#119](https://github.com/zeeps31/purenes/issues/119)) ([`ff43b8c`](https://github.com/zeeps31/purenes/commit/ff43b8cec4069371853f856086d4b5d25263b45e))
* **cpu:** Implement stack instructions ([#118](https://github.com/zeeps31/purenes/issues/118)) ([`8356ace`](https://github.com/zeeps31/purenes/commit/8356ace0cb57ff679a12b0d8a3125c6d67bacd74))
* **cpu:** Implement transfer instructions ([#117](https://github.com/zeeps31/purenes/issues/117)) ([`86888ac`](https://github.com/zeeps31/purenes/commit/86888ac110379b19b5ccddf1e60c08248e218ff5))
* **cpu:** Implement LDY operation ([#116](https://github.com/zeeps31/purenes/issues/116)) ([`c66645b`](https://github.com/zeeps31/purenes/commit/c66645b66ad401f38303cc67112b2f5703bdc7ed))
* **cpu:** Implement LDX operation ([#115](https://github.com/zeeps31/purenes/issues/115)) ([`46c57e6`](https://github.com/zeeps31/purenes/commit/46c57e6592389f0c1d731fbe7285a60efd45a280))
* **cpu:** Implement LDA operation ([#114](https://github.com/zeeps31/purenes/issues/114)) ([`1ad49b0`](https://github.com/zeeps31/purenes/commit/1ad49b06fb2fd5d19c38ffe81965713c03613f79))
* **cpu:** Implement STY operation ([#113](https://github.com/zeeps31/purenes/issues/113)) ([`cb25727`](https://github.com/zeeps31/purenes/commit/cb257272206c332fcc2f6e7d5fecd0a277fdb782))
* **cpu:** Implement STA ([#112](https://github.com/zeeps31/purenes/issues/112)) ([`3012fca`](https://github.com/zeeps31/purenes/commit/3012fca18a969469ac79a4613d841c6560672419))
* **cpu:** Implement ROL operation ([#110](https://github.com/zeeps31/purenes/issues/110)) ([`4ffd880`](https://github.com/zeeps31/purenes/commit/4ffd880ce21b8cd10250454c3001a1943a4178ad))
* **cpu:** Add BIT operation with absolute addressing ([#109](https://github.com/zeeps31/purenes/issues/109)) ([`d45440e`](https://github.com/zeeps31/purenes/commit/d45440ec4d9588ccb2c5eac0ebcd247ed1ed832a))
* **cpu:** Implement BIT operation ([#108](https://github.com/zeeps31/purenes/issues/108)) ([`c419cce`](https://github.com/zeeps31/purenes/commit/c419ccef250ce2e70d11fbee45df9e25374cfd02))
* **cpu:** Implement STX and y-indexed zero-page addressing ([#107](https://github.com/zeeps31/purenes/issues/107)) ([`61f6f4b`](https://github.com/zeeps31/purenes/commit/61f6f4b9e24bb2104b6f5b494d2148a4e5e102aa))
* **cpu:** Implement JMP operation and indirect addressing ([#106](https://github.com/zeeps31/purenes/issues/106)) ([`34700f7`](https://github.com/zeeps31/purenes/commit/34700f781dbf9b6d575171045ad1355ec5ef6600))
* **cpu:** Implement JSR operation ([#105](https://github.com/zeeps31/purenes/issues/105)) ([`779a543`](https://github.com/zeeps31/purenes/commit/779a54388f04dd1528dfaa72a4951fc1eb445ff8))
* **cpu:** Implement AND operation ([#103](https://github.com/zeeps31/purenes/issues/103)) ([`3a1b31c`](https://github.com/zeeps31/purenes/commit/3a1b31ccbfb877aeb0699b8f1dcbc21e4cc22984))
* **cpu:** Implement x-indexed absolute addressing ([#101](https://github.com/zeeps31/purenes/issues/101)) ([`bd63cd6`](https://github.com/zeeps31/purenes/commit/bd63cd6592f5f7daf17118466fac5131fc310a78))
* **cpu:** Implement y-indexed absolute addressing mode ([#99](https://github.com/zeeps31/purenes/issues/99)) ([`c0f4ec8`](https://github.com/zeeps31/purenes/commit/c0f4ec85727651125c808750b9932c2050f11e58))
* **cpu:** Implement flag set operations ([#98](https://github.com/zeeps31/purenes/issues/98)) ([`cd07aee`](https://github.com/zeeps31/purenes/commit/cd07aee772697e95b2763c6d57b4c7acce4c072f))
* **cpu:** Implement flag clear operations ([#97](https://github.com/zeeps31/purenes/issues/97)) ([`66b06b4`](https://github.com/zeeps31/purenes/commit/66b06b4c402b6eb6b99611bd5963193f6b4d6abb))

purenes/palette.py
from typing import Tuple
from typing import List

"""PPU palette defined as RGB tuples.

https://www.nesdev.org/wiki/PPU_palettes
"""
PPU_PALETTE: List[Tuple[int, int, int]] = [
    (84, 84, 84),    (0, 30, 116),    (8, 16, 144),    (48, 0, 136),
    (68, 0, 100),    (92, 0, 48),     (84, 4, 0),      (60, 24, 0),
    (32, 42, 0),     (8, 58, 0),      (0, 64, 0),      (0, 60, 0),
    (0, 50, 60),     (0, 0, 0),       (0, 0, 0),       (0, 0, 0),

    (152, 150, 152), (8, 76, 196),    (48, 50, 236),   (92, 30, 228),
    (136, 20, 176),  (160, 20, 100),  (152, 34, 32),   (120, 60, 0),
    (84, 90, 0),     (40, 114, 0),    (8, 124, 0),     (0, 118, 40),
    (0, 102, 120),   (0,  0, 0),      (0, 0, 0),       (0, 0, 0),

    (236, 238, 236), (76, 154, 236),  (120, 124, 236), (176, 98, 236),
    (228, 84, 236),  (236, 88, 180),  (236, 106, 100), (212, 136, 32),
    (160, 170, 0),   (116, 196, 0),   (76, 208, 32),   (56, 204, 108),
    (56, 180, 204),  (60, 60, 60),    (0, 0, 0),       (0, 0, 0),

    (236, 238, 236), (168, 204, 236), (188, 188, 236), (212, 178, 236),
    (236, 174, 236), (236, 174, 212), (236, 180, 176), (228, 196, 144),
    (204, 210, 120), (180, 222, 120), (168, 226, 144), (152, 226, 180),
    (160, 214, 228), (160, 162, 160), (0, 0, 0),       (0, 0, 0),
]
